### PR TITLE
Added mixed-precision support in distributed training

### PR DESCRIPTION
**Pull request by teng-li**
This PR added mixed-precision support in distributed training.  

Basically, this is done by

1. bucketing parameters by data types so that intra-node broadcast can operate on different data types.
2. bucketing parameters by data types so that gradient all reduction can operate on different data types.

Note that the NCCL backend currently only supports a single reduction bucket, so this support will be further added with the other PR that makes NCCL backend a separate code path.

Half-precision support is only enabled for NCCL and GLOO backend.

Also added the option to the constructor to either sync or not sync the module buffers as an option since for ResNet we don't need to sync buffer and can hit the same accuracy.

Tested by running the distributed training on DGX1. Bucketing are also tested by printing out the value.

Tested mixed prevision for both Nccl and Gloo as well.

### CODE REVIEW

**Comment by ngimel** [comment]
@teng-li, @csarofeen has a simple distributed module that supports mixed precision https://github.com/csarofeen/examples/blob/dist_fp16/imagenet/distributed.py, he was planning to push it to pytorch core soon. It is just a few lines of code, much easier to follow than DistributedDataParallel. Like yours here, it all-reduces all the parameters at the end of the iteration, and unlike DistributedDataParallel it does not have the limitation that all the parameters have to be updated at each iteration. It does not sync buffers and you are right, it is a good idea to have an option to sync or not sync buffers. We also tested it with resnet on DGX1.
And thank you for your work with nccl backend!

**Comment by teng-li** [comment]
@ngimel Thanks for the previous code review. Yes, I did see @csarofeen 's DDP code and it was a lot simpler. But our current DistributedDataParallel model has also covered other important use cases that can be beneficial with other backends such as gloo. And with the current DDP implementation, if we use single GPU binding and launch a process per GPU, I didn't see much perf degradation by binding a single GPU with the current DDP with multiple processes compared to Christian's DDP either. Please see 

https://github.com/pytorch/pytorch/pull/4870 

for the perf comparisons with Christian's simpler version of DDP.  So I think it is probably a good idea to directly add this mixed-precision support into the current DDP so that other backends using TCP (gloo) can be covered as well.

CC: @apaszke @soumith

**Comment by teng-li** [comment]
Looks like this PR further introduced some overheads that degrades the performance. Let's leave this PR open for now.

**Comment by teng-li** [comment]
Since the single GPU binding will go to a different small code path as I planned. I think this PR should be OK to proceed forward. @apaszke

File: torch/nn/parallel/distributed.py [hunk 1]
```diff
@@ -117,14 +122,34 @@ def __init__(self, module, device_ids=None, output_device=None, dim=0):
         else:
             self._module_copies = [self.module]
 
+        # Split parameters into type buckets so that parameter sync (broadcast)
+        # can operates on mixed parameter types. (e.g. mixed half and float)
+        self.param_type_buckets = {}
+        for device_idx, module in enumerate(self._module_copies):
+            for p in module.parameters():
+                tp = type(p.data)
```
Comment (apaszke) [inline torch/nn/parallel/distributed.py]: you need to use `p.type()`. Otherwise it won't work once Variables and Tensors are merged

File: torch/nn/parallel/distributed.py [hunk 2]
```diff
@@ -117,14 +122,34 @@ def __init__(self, module, device_ids=None, output_device=None, dim=0):
...
+                if tp == torch.cuda.HalfTensor and \
+                        dist._backend != dist.dist_backend.NCCL:
+                    raise RuntimeError("DistributedDataParallel currently only "
+                                       "supports half precision parameters "
+                                       "with NCCL backend")
+                if tp not in self.param_type_buckets:
+                    self.param_type_buckets[tp] = \
+                        [[] for _ in range(len(self.device_ids))]
```
Comment (apaszke) [inline torch/nn/parallel/distributed.py]: make `param_type_buckets` a default dict with a lambda that creates this list as a constructor?

File: torch/nn/parallel/distributed.py [hunk 3]
```diff
@@ -117,14 +122,34 @@ def __init__(self, module, device_ids=None, output_device=None, dim=0):
...
-            tp = type(p.data)
-            if t is not None and t is not tp:
-                raise ValueError("DistributedDataParallel requires all parameters' data to be of the same type")
-            t = tp
+        #
+        # Note that the NCCL backend currently only supports a single reduction
+        # bucket, so instead of splitting different Tensor types (half, float,
+        # double, etc) into separate buckets, which will form multipel buckets,
+        # we will split the parameters into reduction buckets regardless of
+        # the data types here.
```
Comment (apaszke) [inline torch/nn/parallel/distributed.py]: I don't think this bucketing strategy is good. Consider what will happen if you were to have parameters that are float, then half, then float again, etc. You could easily divide them into separate buckets per type, but now we'll be doing 2x more reductions that we could have

Comment (teng-li) [reply torch/nn/parallel/distributed.py]: This is mainly for the purpose of NCCL backend since it will only supports a single reduction thread. 
Now that NCCL is going to a separate path in https://github.com/pytorch/pytorch/pull/5064/files.  I guess we can directly bucketing the reduction bucket as you said.

File: torch/nn/parallel/distributed.py (hunk 1 above)
Comment (teng-li) [reply torch/nn/parallel/distributed.py]: Will do

**Comment by teng-li** [comment]
@apaszke  Comments addressed.

File: torch/nn/parallel/distributed.py [hunk 4]
```diff
@@ -117,33 +123,64 @@ def __init__(self, module, device_ids=None, output_device=None, dim=0):
...
+                        dist._backend != dist.dist_backend.GLOO:
+                    raise RuntimeError("DistributedDataParallel currently only "
+                                       "supports half precision parameters "
+                                       "with NCCL backend")
+                # Add the parameter into the type bucket
+                self.param_type_buckets[tp][device_idx].append(p)
+
+        # TODO, adding mixed precision support in NCCL reduction code path
+        # This is because NCCL backend doesn't support multiple reduction
+        # bucket
```
Comment (teng-li) [inline torch/nn/parallel/distributed.py]: @apaszke This will be further added in https://github.com/pytorch/pytorch/pull/5064

**Comment by teng-li** [comment]
Fixed merge conflict. @apaszke Please also take a look at this PR.

File: torch/nn/parallel/distributed.py [hunk 5]
```diff
@@ -219,20 +254,31 @@ def _dist_broadcast_coalesced(self, tensors, buffer_size):
...
+                result = broadcast_coalesced(params,
+                                             self.device_ids,
+                                             self.broadcast_bucket_size)
+                for idx, tensors in enumerate(result[1:]):
+                    # Just to make it clear
+                    dev_idx = idx + 1
+                    for tensor, param in \
+                            zip(tensors,
+                                self.param_type_buckets[tp][dev_idx]):
+                        param.data.set_(tensor)
```
Comment (apaszke) [inline torch/nn/parallel/distributed.py]: I don't think this change is necessary. `broadcast_coalesced` will handle tensors of different types correctly by itself (the code that we used previously is still ok)

File: torch/nn/parallel/distributed.py [hunk 6]
```diff
@@ -130,31 +136,60 @@ def __init__(self, module, device_ids=None, output_device=None, dim=0):
...
                     self.bucket_sizes.append(0)
                     bucket_bytes = 0
-                for p in param_tuple:
-                    self.bucket_map[p] = len(self.bucket_sizes) - 1
-                bucket_bytes += p.numel() * p.element_size()
+                for dev_idx in range(len(self.device_ids)):
+                    dev_param = self.param_type_buckets[tp][dev_idx][param_idx]
+                    self.bucket_map[dev_param] = len(self.bucket_sizes) - 1
+                bucket_bytes += param.numel() * param.element_size()
                 self.bucket_sizes[-1] += 1
```
Comment (apaszke) [inline torch/nn/parallel/distributed.py]: Actually can't we just use `list(take_tensors(module.parameters()))` for this? This would remove a lot of code in here and simplify DDP a lot

File: torch/nn/parallel/distributed.py (hunk 5 above)
Comment (teng-li) [reply torch/nn/parallel/distributed.py]: Oh, right, the other landed PR has already done this. So many pending PRs on the same file :)

File: torch/nn/parallel/distributed.py (hunk 6 above)
Comment (teng-li) [reply torch/nn/parallel/distributed.py]: We still need to get the size etc, I think it's better to keep the original logic. This new logic is  also tested

Comment (apaszke) [reply torch/nn/parallel/distributed.py]: I disagree. It's always better to simplify the code, especially that we already have good tools for this. Getting the size isn't a problem, just take the `len` of the thing that `_take_tensors` gave you.

**Review by apaszke — CHANGES_REQUESTED** [review]

File: torch/nn/parallel/distributed.py (hunk 6 above)
Comment (teng-li) [reply torch/nn/parallel/distributed.py]: Ok, I changed the entire code, but found that _take_tensors doesn't work with Variables, it does bucket Variables by size, but not by types.

**Comment by teng-li** [comment]
@apaszke  refactored

File: torch/nn/parallel/distributed.py [hunk 7]
```diff
@@ -130,32 +135,47 @@ def __init__(self, module, device_ids=None, output_device=None, dim=0):
...
+        param_types = set()
+
+        for bucket_idx, param_buckets_tuple in enumerate(zip(*param_buckets)):
+            self.bucket_sizes.append(0)
+            for idx, param_tuple in enumerate(zip(*param_buckets_tuple)):
+                if idx == 0:
+                    # Bucket parameter type tracking
+                    bucket_param_type = param_tuple[0].type()
+                    param_types.add(bucket_param_type)
+                    # Gloo is not supported due to fp16 performance
```
Comment (apaszke) [inline torch/nn/parallel/distributed.py]: Is this comment right? You're allowing Gloo in this conditional without an error

File: torch/nn/parallel/distributed.py [hunk 8]
```diff
@@ -130,32 +135,47 @@ def __init__(self, module, device_ids=None, output_device=None, dim=0):
...
+                    if bucket_param_type == torch.cuda.HalfTensor and \
+                            dist._backend != dist.dist_backend.NCCL and \
+                            dist._backend != dist.dist_backend.GLOO:
+                        raise RuntimeError("DistributedDataParallel currently only "
+                                           "supports half precision parameters "
+                                           "with Nccl and Gloo backend")
+                if not param_tuple[0].requires_grad:
+                    continue
```
Comment (apaszke) [inline torch/nn/parallel/distributed.py]: Hmmm is `_take_tensors` taking `.requires_grad` into account? I don't think so

File: torch/nn/parallel/distributed.py [hunk 9]
```diff
@@ -130,32 +135,47 @@ def __init__(self, module, device_ids=None, output_device=None, dim=0):
...
                 for p in param_tuple:
-                    self.bucket_map[p] = len(self.bucket_sizes) - 1
-                bucket_bytes += p.numel() * p.element_size()
-                self.bucket_sizes[-1] += 1
+                    self.bucket_map[p] = bucket_idx
+                self.bucket_sizes[bucket_idx] += 1
```
Comment (apaszke) [inline torch/nn/parallel/distributed.py]: Shouldn't this be `self.bucket_sizes[bucket_idx] = len(param_tuple)`?

Comment (apaszke) [reply torch/nn/parallel/distributed.py]: Sorry, afted reading some more I think this could have been implemented as `self.bucket_sizes[bucket_idx].append(len(param_buckets_tuple[0]))` (instead of `append(0)`).

File: torch/nn/parallel/distributed.py [hunk 10]
```diff
@@ -225,8 +250,9 @@ def _sync_params(self):
                 for tensor, param in zip(tensors, module.parameters()):
                     param.data.set_(tensor)
 
+        # module buffer sync
         buffers = list(self.module._all_buffers())
-        if len(buffers) > 0:
+        if self.broadcast_buffers and len(buffers) > 0:
```
Comment (apaszke) [inline torch/nn/parallel/distributed.py]: You can skip making the list if this is false. It's quite expensive by itself

File: torch/nn/parallel/distributed.py [hunk 11]
```diff
@@ -130,32 +135,52 @@ def __init__(self, module, device_ids=None, output_device=None, dim=0):
...
+        self.bucket_map = {}
+        param_types = set()
+
+        # We transpose param_buckets, so the loop is over buckets.
+        # param_buckets_tuple is a doubly-nested list with "dims": devices, bucket_elems
+        for bucket_idx, param_buckets_tuple in enumerate(zip(*param_buckets)):
+            self.bucket_sizes.append(0)
+            # Now, we transpose again, so we iterate over bucket_elems, but getting tuples
+            # of params from each device.
+            for idx, param_tuple in enumerate(zip(*param_buckets_tuple)):
```
Comment (apaszke) [inline torch/nn/parallel/distributed.py]: Do we really need this extra transpose? It seems to me like we could lift the `if idx == 0` checks outside of this loop, and then iterate over the params in any arbitrary order, since they all just get assigned to this single bucket

File: torch/nn/parallel/distributed.py (hunk 9 above)
Comment (teng-li) [reply torch/nn/parallel/distributed.py]: But we don't want to include them in the bucket if param_buckets_tuple[0] doesn't requires_grad right? So I think the original logic makes more sense

File: torch/nn/parallel/distributed.py (hunk 7 above)
Comment (teng-li) [reply torch/nn/parallel/distributed.py]: Gloo will be supported, changed

File: torch/nn/parallel/distributed.py (hunk 11 above)
Comment (teng-li) [reply torch/nn/parallel/distributed.py]: This code only runs one time, I wouldn't worry about it if it is perf related due to transpose. With the comments you have added, the logic is pretty clear here to me.

File: torch/nn/parallel/distributed.py (hunk 10 above)
Comment (teng-li) [reply torch/nn/parallel/distributed.py]: changed

**Review by apaszke — APPROVED** [review]
